demos/sslecho/main.c

```diff
@@ -0,0 +1,301 @@
+/*
+ *  Copyright 20xx-20yy The OpenSSL Project Authors. All Rights Reserved.
+ *
+ *  Licensed under the Apache License 2.0 (the "License").  You may not use
+ *  this file except in compliance with the License.  You can obtain a copy
+ *  in the file LICENSE in the source distribution or at
+ *  https://www.openssl.org/source/license.html
+ */
+
+#include <stdio.h>
+#include <unistd.h>
+#include <string.h>
+#include <sys/socket.h>
+#include <arpa/inet.h>
+#include <openssl/ssl.h>
+#include <openssl/err.h>
+
+static const char		*revision = "1.0.3";
+
+static const int        server_port = 4433;
+
+typedef unsigned char	bool;
+#define	true			1
+#define	false			0
+
+/*
+ * This flag won't be useful until both accept/read (TCP & SSL) methods
+ * can be called with a timeout. TBD.
+ */
+static volatile bool    server_running = true;
+
+int create_socket(bool isServer) {
+	int 				s;
+	int					optval = 1;
+	struct sockaddr_in	addr;
+
+	s = socket(AF_INET, SOCK_STREAM, 0);
+	if (s < 0) {
+		perror("Unable to create socket");
+		exit(EXIT_FAILURE);
+	}
+
+	if (isServer) {
+		addr.sin_family = AF_INET;
+		addr.sin_port = htons(server_port);
+		addr.sin_addr.s_addr = INADDR_ANY;
+
+		/* Reuse the address; good for quick restarts */
+		if (setsockopt(s, SOL_SOCKET, SO_REUSEADDR, &optval, sizeof(optval))
+				< 0) {
+			perror("setsockopt(SO_REUSEADDR) failed");
+			exit(EXIT_FAILURE);
+		}
+
+		if (bind(s, (struct sockaddr*) &addr, sizeof(addr)) < 0) {
+			perror("Unable to bind");
+			exit(EXIT_FAILURE);
+		}
+
+		if (listen(s, 1) < 0) {
+			perror("Unable to listen");
+			exit(EXIT_FAILURE);
+		}
+	}
+
+	return s;
+}
+
+SSL_CTX* create_context(bool isServer) {
+	const SSL_METHOD *method;
+	SSL_CTX *ctx;
+
+	if (isServer)
+		method = TLS_server_method();
+	else
+		method = TLS_client_method();
+
+	ctx = SSL_CTX_new(method);
+	if (!ctx) {
+		perror("Unable to create SSL context");
+		ERR_print_errors_fp(stderr);
+		exit(EXIT_FAILURE);
+	}
+
+	return ctx;
+}
+
+void configure_context(SSL_CTX *ctx) {
+	/* Set the key and cert */
+	if (SSL_CTX_use_certificate_file(ctx, "cert.pem", SSL_FILETYPE_PEM) <= 0) {
+		ERR_print_errors_fp(stderr);
+		exit(EXIT_FAILURE);
+	}
+
+	if (SSL_CTX_use_PrivateKey_file(ctx, "key.pem", SSL_FILETYPE_PEM) <= 0) {
```

Comment:
The use certificate and PrivateKey function calls are really only necessary for server side use in this simple example.

Additionally I think there are some calls missing on the client side:

First:

````
		SSL_CTX_set_verify(ctx, SSL_VERIFY_PEER, NULL);
````
This will cause the client to immediately abort the connection in the event that a certificate verify fails. Without this clients need to call `SSL_get_verify_result()` to check the handshake verification result manually.

Secondly, we should probably do this:
````
		SSL_CTX_load_verify_file(ctx, "cert.pem");
````
This tells the client it can trust the server's self signed cert. In a real application you might call `SSL_CTX_load_verify_dir` to specify the location of the trust store (or one of the other similar functions defined on the same man page). But this should be good enough for this simple app.

Reply:
The SSL_CTX_load_verify_file() call does not compile.
Function name not in docs, and not found in headers?
I added it to the code with a comment to that effect...

Reply:
It's in the docs here:
https://www.openssl.org/docs/manmaster/man3/SSL_CTX_load_verify_file.html

In the headers here (in ssl.h.in which gets translated to ssl.h when OpenSSL is built):
https://github.com/openssl/openssl/blob/dd2fcc1f7c44c5fb5aa2d33aecdc699c7018ce01/include/openssl/ssl.h.in#L2051

Reply:
These are new in openssl-3.0

Reply:
Ah..yes...in 1.1.1 and earlier you had to instead use `SSL_CTX_load_verify_locations` - but since this demo will be going into master its fine to use `SSL_CTX_load_verify_file`

Reply:
SSL_CTX_load_verify_dir not found either.
Additionally, the SSL_CTX_set_verify() causes the client to fail.
A call to SSL_get_verify_result() after a "successful" connection also causes the client to fail.
I'm surely doing this wrong, but don't know why.
Will commit the requested changes that don't break it.